mypy/messages.py

```diff
@@ -992,6 +997,28 @@ def pretty_or(args: List[str]) -> str:
     return ", ".join(quoted[:-1]) + ", or " + quoted[-1]
 
 
+def invariance_notes(notes: List[str], arg_type: Type, expected_type: Type) -> List[str]:
+    """Explain that the type is invariant and give notes for how to solve the issue."""
+    if isinstance(arg_type, Instance) and isinstance(expected_type, Instance):
+        if (arg_type.type.fullname() == 'builtins.list' and
+                expected_type.type.fullname() == 'builtins.list'):
+            invariant_type = 'List'
+            covariant_suggestion = 'Consider using "Sequence" instead, which is covariant'
+        elif (arg_type.type.fullname() == 'builtins.dict' and
+              expected_type.type.fullname() == 'builtins.dict' and
+              arg_type.args[0].type == expected_type.args[0].type and
+              arg_type.args[1].type != expected_type.args[1].type):
```

Comment:
There is something important here missing (also for ``List``): we should show the note only when using a covariant type will actually help, for example here:
```python
x: List[int]
y: List[str]
x = y
```
we don't need to show any notes, since using ``Sequence`` will not help. So that instead ``arg_type.args[1].type != expected_type.args[1].type`` use ``if subtypes.is_subtype(arg_type.args[1], expected_type.args[1])``.